heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+
+### Batch scripts
+
+To make things easier, you can write all the code above into a single
+`.saw` file and process it in a batch. The file `ffs.saw` would look like this:
+
+```bash
+print "Loading module ffs.bc";
+l <- llvm_load_module "ffs.bc";
+
+print "Extracting reference term: ffs_ref";
+ffs_ref <- llvm_extract l "ffs_ref";
+
+print "Extracting implementation term: ffs_imp";
+ffs_imp <- llvm_extract l "ffs_imp";
+
+print "Proving equivalence: ffs_ref == ffs_imp";
+let thm1 = {{ \x -> ffs_ref x == ffs_imp x }};
+result <- prove z3 thm1;
+print result;
+
+print "Done."
+```
+
+If you save the file in the same directory you can run:
+```bash
+% cabal run saw -- ffs.saw
+Up to date
+
+[16:49:13.646] Loading file "/PATH/TO/SAW/saw-script/doc/tutorial/code/my_ffs.saw"
+[16:49:13.647] Loading module ffs.bc
+[16:49:13.651] Extracting reference term: ffs_ref
+[16:49:13.663] Extracting implementation term: ffs_imp
+[16:49:13.666] Proving equivalence: ffs_ref == ffs_imp
+[16:49:13.716] Valid
+[16:49:13.716] Done.
+```
+
+That's it! You know the basics of SAW.
+
+## Using Heapster
+
+Heapster is, fundamentally, a type system for extracting functional
+specifications from memory-safe imperative programs. The type system,
+defined inside SAW, uses separation types to reason about memory
+safety. Once a program is type-checked as memory-safe, it can be then
+extracted as a functional program to be verified in Coq.
+
+**TODO: Double check this description of Heapster**
+
+### Heapster type-checking overview
+
+Heapster allows us to (1) type check programs with respect to
+types/specifications that can express separation loigc and (2) extract
+the resulting functional program to Coq for further verification.
+
+The process will generally involve
+
+- [1. Generating LLVM bitcode](#1-generating-llvm-bitcode)
+- [2. Run the SAW interpreter with Heapster](#2-run-the-saw-interpreter-with-heapster)
+- [3. Load the file and extract the two function specifications.](#3-load-the-file-and-extract-the-two-function-specifications-1)
+- [4. Writing heapster specifications for your functions](#4-writing-heapster-specifications-for-your-functions)
+- [5. Writing a SAW script to type-check your code with respect to the sepcification](#5-writing-a-saw-script-to-type-check-your-code-with-respect-to-the-sepcification)
+- [6. Writing a Coq file to prove things about the generated functional specification(s)](#6-writing-a-coq-file-to-prove-things-about-the-generated-functional-specifications)
+
+Just like with SAW, Heapster can be processed in batch. To do so, you
+can combine steps 2-6 in a `.saw` file and use SAW's batch processing.
+
+### Running an example
+
+This section will walk through the process of using Heapster to write
+and verify some C code. Specifially, we want to verify the function
+`is_elem`, which tests if a specific value is in a list. The function,
+together with others, can be found in `examples/linked_list.c`.
```

Comment:
Instead of expecting the user to be in the `heapster-saw` directory for all this tutorial, you should have them be in `heapster-saw/examples`, so that the user can access the Makefile listed there. Maybe put an explicit `cd heapster-saw/examples` command in section 1 (or 0)?

Reply:
If it makes sense, I'll add it at the beginning of the Heapster section, since installation and running Saw makes more sense from the top level directory.